Initial base for AltFP (#2)
riscv/riscv-alt-fp: doc/riscv-bfloat16-format.adoc

[[bfloat16_format]]
== BFloat16 Operand Format

BFloat16 bits::
[wavedrom, , svg]
....
{reg:[
{bits: 1, name: 'S'},
{bits: 5, name: 'expo'},
{bits: 10, name: 'frac'},
]}
....

IEEE Compliance: While BFloat16 (BF16) is not an IEEE-754 _standard_ format, it is a valid floating point format as defined by the standard. There are three parameters that specify a format: radix (b), number of digits in the significand (p), and maximum exponent (emax).
For BF16 these values are:

[%autowidth]
.BFloat16 paramenters
|===
|radix (b)|2
|significand (p)|8
|emax|127
|===

.Obligatory Floating Point Format Table
[cols = "1,1,1,1,1,1,1,1"]
|===
|Format|Sign Bits|Expo Bits|fraction bits|padded 0s|encoding bits|expo max/bias|expo min

|FP16    |1| 8| 7| 0|16| 127|-126
|BFloat16|1| 5|10| 0|16|  15| -14
|TF32    |1| 8|10|13|32| 127|-126
|FP32    |1| 9|23| 0|32| 127|-126
|FP64    |1|11|52| 0|64|1023|-1022
|FP128   |1|15|112|0|128|16,383|-16,382
|===

== BFloat16 behaviors

=== Subnormal Numbers:

> Floating-point values that are too small to represnted as normal numbers, but can still be represnted by using the format's smallest exponent with a zero integer bit and one or more leading 0s --- and one or

Reviewer: ```suggestion
Floating-point values that are too small to be represented as normal numbers, but can still be represented by using the format's smallest exponent with a zero integer bit and one or more leading 0s --- and one or

```

more 1s --- in the trailing fractional bits are called subnormal numbers. Basically, the idea is there is
a trade off of precision to support _gradual underflow_.

> In RISC-V BFloat16, all subnormal BFloat16 inputs as treated as zero and subnormal outputs are flushed to zero. The sign of the original value is retained. This is not consistent with '754' but has been found to be a suitable alternative in many workloads. Furthermore, with BFloat16's relatively large exponent range, subnormals add little value.

Reviewer: ```suggestion
In RISC-V BFloat16, all subnormal BFloat16 inputs are treated as zero and subnormal outputs are flushed to zero. The sign of the original value is retained. This is not consistent with '754' but has been found to be a suitable alternative in many workloads. Furthermore, with BFloat16's relatively large exponent range, subnormals add little value.

```

 
===  Infinities:

> Infinties are used to represnt values that are too large to be represnted by the target format. These are usually produced as a result of overflows (depending on the rounding mode), but can also be provided as inputs. Infities have a sign associated with them: there are positive infionioties and negative infinities.

Reviewer: ```suggestion
Infinities are used to represent values that are too large to be represented by the target format. These are usually produced as a result of overflows (depending on the rounding mode), but can also be provided as inputs. Infinities have a sign associated with them: there are positive infinities and negative infinities.

```

[[bfloat16_format]]
== BFloat16 Operand Format

BFloat16 bits::
[wavedrom, , svg]
....
{reg:[
{bits: 1, name: 'S'},
{bits: 5, name: 'expo'},
{bits: 10, name: 'frac'},
]}
....

IEEE Compliance: While BFloat16 (BF16) is not an IEEE-754 _standard_ format, it is a valid floating point format as defined by the standard. There are three parameters that specify a format: radix (b), number of digits in the significand (p), and maximum exponent (emax).
For BF16 these values are:

[%autowidth]
.BFloat16 paramenters
|===
|radix (b)|2
|significand (p)|8
|emax|127
|===

.Obligatory Floating Point Format Table
[cols = "1,1,1,1,1,1,1,1"]
|===
|Format|Sign Bits|Expo Bits|fraction bits|padded 0s|encoding bits|expo max/bias|expo min

|FP16    |1| 8| 7| 0|16| 127|-126

> |BFloat16|1| 5|10| 0|16|  15| -14

Reviewer 2: I think you got these switched around: BF16 is the same as FP32, but with 16 fewer fraction bits: https://cloud.google.com/blog/products/ai-machine-learning/bfloat16-the-secret-to-high-performance-on-cloud-tpus

Author: You are correct. Somehow I swapped them.

|TF32    |1| 8|10|13|32| 127|-126

> |FP32    |1| 9|23| 0|32| 127|-126

Reviewer 2: FP32 only has 8 exponent bits, not 9 - as it's written now the sum of bits would be 33.
There _is_ an implied `1` bit in there so technically you get the effect of 33 bits, but that goes into the fraction 🙂

Author: Thanks for catching the typo. I will fix.

[[bfloat16_format]]
== BFloat16 Operand Format

BFloat16 bits::
[wavedrom, , svg]
....
{reg:[
{bits: 1, name: 'S'},
{bits: 5, name: 'expo'},
{bits: 10, name: 'frac'},
]}
....

IEEE Compliance: While BFloat16 (BF16) is not an IEEE-754 _standard_ format, it is a valid floating point format as defined by the standard. There are three parameters that specify a format: radix (b), number of digits in the significand (p), and maximum exponent (emax).
For BF16 these values are:

[%autowidth]
.BFloat16 parameters
|===
|radix (b)|2
|significand (p)|8
|emax|127
|===

.Obligatory Floating Point Format Table
[cols = "1,1,1,1,1,1,1,1"]
|===
|Format|Sign Bits|Expo Bits|fraction bits|padded 0s|encoding bits|expo max/bias|expo min

|FP16    |1| 8| 7| 0|16| 127|-126
|BFloat16|1| 5|10| 0|16|  15| -14
|TF32    |1| 8|10|13|32| 127|-126
|FP32    |1| 8|23| 0|32| 127|-126
|FP64    |1|11|52| 0|64|1023|-1022
|FP128   |1|15|112|0|128|16,383|-16,382
|===

== BFloat16 behaviors

=== Subnormal Numbers:

> Floating-point values that are too small to be represented as normal numbers, but can still be represented by using the format's smallest exponent with a zero integer bit and one or more leading 0s --- and one or
> 
> more 1s --- in the trailing fractional bits are called subnormal numbers. Basically, the idea is there is

Reviewer: ```suggestion
Floating-point values that are too small to be represented as normal numbers, but can still be represented by using the format's smallest exponent with a zero integer bit and one or more leading 0s --- and one or 
more 1s --- in the trailing fractional bits are called subnormal numbers. Basically, the idea is there is

```

Author: Thanks for catching. I fixed this but hadn't added it before committing. The latest pull request should look better (for this file anyway).

Reviewer: No problem, I am trying to do a full review and listing typos / questions along the way.

a trade off of precision to support _gradual underflow_.

In RISC-V instructions operating on BFloat16, it is generally intended that all subnormal BFloat16 inputs are treated as zero and subnormal outputs are flushed to zero. The sign of the original value is retained. However, it

> is uop to the instruction to specify this behavior. 

Reviewer: ```suggestion
is up to the instruction to specify this behavior. 

```

> vary based on the instruction as there are special cases where it may be undesirable to 
> some special cases where it is not desirable to treat 
> This is not consistent with '754' but has been found to be a suitable alternative in many workloads. Furthermore, with BFloat16's relatively large exponent range, subnormals add little value.

Reviewer: ```suggestion
This behavior may vary based on the instruction as there are special cases where it may be undesirable to treat subnormals as zero.
This is not consistent with '754' but has been found to be a suitable alternative in many workloads. Furthermore, with BFloat16's relatively large exponent range, subnormals add little value.

```

as a number or infinity. For example, performing the square root of -1 will result in a NaN because
there is no real number that can represent the result. NaNs can also be used as inputs.

There are two types of NaNs: signalling and quiet. Signalling NaNs are provided as input data since no computational instruction will ever produce this kind of a NaN. Operating on a Signalling NaN will produce an invalid operation exception. Operating on a Quiet NaN usually does not cause an exception.

NaNs include a sign bit, but the bit has no meaning.

NaNs are important for keeping meaningless results from being operated upon. It is best to retain them. As IEEE allows, operations should return the canonical NaN rather than be required to propagate the payload.
 
===  Rounding Modes:
In general, the default IEEE rounding mode (round to nearest, ties to even) works for arithmetic cases. There are some special cases where a particular instruction benefits from a different rounding mode (e.g., convert to integer, widening multiply-accumulate) - we can address this on those specific instructions.

>  

Reviewer: Does this mean we intend to have a static rounding mode force to RNE by default and only allow rounding-mode static (opcode) or dynamic (csr) selection on a specific subset of instructions ? This seems to be in contradiction with F and D extensions and should be justified here IMHO.

Author: Yes, these instructions would have a static rounding mode that is not overridable. Yes, this is different from '754. However, it is a common simplification (just like flushing subnormals). If someone needs more control over the rounding mode they can run in SP (F).

I agree that we will need to provide a detailed justification in the specification for this simplification.

=== Handling exceptions

> Default exception handling, as defined by IEEE, is a simple and effective approach to producing results in exceptional cases. For the coder to be able to see what has happened, and take further action if needed, the BFloat16 instructions need to set floating-point exception flags the same way as all other floating-point instructions in RISC-V. 

Reviewer: This formulation may not be future proof, we may want to cite explicitly the basic floating-point extensions here.

Author: Which area are concerned about: the rounding mode, default exception handling, or both?

Should the need arise, an extension could be added that allows the rounding mode to be changed by the CSR.

The handling of exceptions via the IEEE default is common across RISC-V. Is this what you mean about citing  the basic floating-point extensions?
At some point there might be a TG that creates trapped exceptions for FP instructions - but right now only _default_ is supported.

Reviewer: I am concerned that another floating-point extension may be introduced with a different way of managing FP exception flags, making "as all other floating-point instructions ..." missleading. So mentioning explicitly that we intend to managed them as extension F and D (and Q), clarifies things if such an extension should appear at some point. I agree that the use  of such remark may be limited.
